### DISCUSSION TITLE
Getting Wrong Matrix form of Operator during implementation of six-fold rotational symmetry

### DISCUSSION THREAD

**Discussion by DHIMAN002** [QuSpin/QuSpin#507] [Q&A] [answered]
I have started using QuSpin and I got some problem while using it to construct a simple Hermitian operator for the following simple system. I have a small system with system size 7 with six-fold rotational symmetry as shown in the attached figure (see at the bottom). Next I try to calculate the Sz-operator for each site. The outcome is correct without application of six-fold symmetry. But the outcome is wrong when I implemented six fold rotation,( I compared with the derivation using pen and paper.). I have focused only for one particle sector and $\nu=0$ rotational quantum number. Most of the code is same as in Quspin Tutorial  https://weinbe58.github.io/QuSpin/examples/user-basis_example0.html#user-basis-example0-label, except the symmetry maps.

### Here is some part of the code in QuSpin:

_Defining the symmetry operator (I used  http://programming.sirrida.de/calcperm.php to generate the symmetry map) and the expected outcome of it [0,1,2,3,4,5,6]--> [5,0,1,2,3,4,6], (QuSpin convention wise [0,1,2,3,4,5,6]-->[0,6,1,2,3,4,5], because it considers the highest bit as particle at 0-th site, 6 becomes 0)_

```

@cfunc(map_sig_32,
       locals=dict())
def Rotation(x,N,sign_ptr,args):
    return ((x & 0x00000040)
            | ((x & 0x00000001) << 5)
            | ((x & 0x0000003e) >> 1))


R_args=np.array([],dtype=np.uint32)
maps =dict(R_block=(Rotation,6,0,R_args))
basis = user_basis(np.uint32,N,op_dict,allowed_ops=set("+-xyznI"),sps=2,pcon_dict=pcon_dict,**maps)
```

_The other dictionaries like pcon_dict, op_dict is same as in https://weinbe58.github.io/QuSpin/examples/user-basis_example0.html#user-basis-example0-label for a spin-half Heisenberg chain, there is no change._

_Here is the code for calculating the Sz-operator for each site,_
```
for i in range(7):
  print("Site=",i)
  static=[["z",[[1,i]]]]
  dynamic=[]
  no_checks=dict(check_symm=False, check_pcon=False, check_herm=False)
  Szi=hamiltonian(static,[],basis=basis,dtype=np.float64,**no_checks)
  print(Szi)

```

### The outcome from QuSpin:
_I get the following outcome,_

Site= 0
static mat: 
  (0, 0)        1.0
  (1, 1)        -1.0

dynamic:

Site= 1
static mat: 
  (0, 0)        -1.0
  (1, 1)        1.0

dynamic:

Site= 2
static mat: 
  (0, 0)        -1.0
  (1, 1)        -1.0

dynamic:

Site= 3
static mat: 
  (0, 0)        -1.0
  (1, 1)        -1.0

dynamic:

Site= 4
static mat: 
  (0, 0)        -1.0
  (1, 1)        -1.0

dynamic:

Site= 5
static mat: 
  (0, 0)        -1.0
  (1, 1)        -1.0

dynamic:

Site= 6
static mat: 
  (0, 0)        -1.0
  (1, 1)        -1.0

dynamic:

### The Expected Outcome (Pen and paper calculation):
Firstly due to six-fold symmetry the Sz operators on sites 1,2,3,4,5,6 should be same for $\nu=0$ sector. The basis for \nu=0 sector with one particle should be,

|0>= |1000000>
|1>=\frac{1}{\sqrt{6}} (|0100000>+|0010000>+|0001000>+|0000100>+|0000010>+|0000001>)

The corresponding Sz in the matrix form for site-0 should be,
Sz=[[1,0],[0,-1]]
The operator Sz for sites 1,2,3,4,5,6 should be,
Sz=[[-1,0],[0,-2/3]]

But the QuSpin outcome of the operators at site 1,2,3,4,5,6 are different. 

![SpinHalfSystem](https://user-images.githubusercontent.com/67744770/156490416-5c964f0e-7068-479f-a46e-d22765485650.jpg)

**Comment by mgbukov** [comment]
it looks like your operator Sz_i, defined using `static=[["z",[[1,i]]]]` is not invariant under the symmetry so it should not have a well-defined quantum number under this symmetry. Try summing up sites 1 thru 7 to make it invariant. 

If this doesn't help, remove the central site and try to see if you can reproduce the 6-fold rotation symmetry of the system only; here you can easily benchmark against `spin_basis_1d` with simple translation symmetry (i.e. 6 sites with periodic boundary condition).

**Comment by DHIMAN002** [comment]
Hi Marin,
Thanks for prompt reply. 
I want to calculate average of Spin z-component at each site. So how to do it ?

I thought it possible to find the average of operator in a basis with a quantum number representing a symmetry, even if the operator does not follow the symmetry.
For example we know that `<Sx>=0` and `<Sy>=0,` in the basis where the quantum number is `Sz`, although `Sx` and `Sy` are not the operator with good quantum number.

**Comment by mgbukov** [comment, marked as answer]
> I want to calculate average of Spin z-component at each site. So how to do it ?

If you have the symmetry, then the value of Sz will be the same on sites 1-7, so you can use $1/6\sum_{j=1}^7 S^z_j$ which is invariant under the symmetry; to get the value at site 0, you can compute Sz_0 separately. 

> I thought it possible to find the average of operator in a basis with a quantum number representing a symmetry, even if the operator does not follow the symmetry.
For example we know that <Sx>=0 and <Sy>=0, in the basis where the quantum number is Sz, although Sx and Sy are not the operator with good quantum number.

It doesn't work this way, unfortunately; Working in a symmetry sector means that you block diagonalize your operator first and then you apply a projector to the given sector. If your operator does not become block diagonal, then the projection will change it.

**Reply by DHIMAN002** [reply]
Thank you Marin, it clarified my confusion.
